### DISCUSSION TITLE
Federated, pluggable upstream OAuth/OIDC providers

### DISCUSSION THREAD

**Discussion by coopbri** [kanidm/kanidm#3163] [Q&A] [answered]
There might be information about this in the docs or an open discussion already, but I couldn't find -- does kanidm support (or is planning to support) plugging into existing upstream OAuth2 providers?

Prior art:
- https://github.com/dexidp/dex (this is what DEX does at its core, federated gateway for upstream auth)
- https://www.keycloak.org (see "Identity Brokering and Social Login" section on this landing page)

Of course, I could use a project similar to the above to solve this use case, but I am curious if kanidm itself supports/is planning to support it natively.

**Comment by Firstyear** [comment]
What are you trying to achieve?

**Reply by coopbri** [reply]
Wow! Fast response. I have a mix of in-house OIDC providers I support, and just want a single gateway to connect my downstream applications to them (app -> gateway -> one of multiple upstream OIDC providers). Basically, what dex does exactly.

**Reply by Firstyear** [reply, marked as answer]
Right, so it sounds like you have a bunch of oidc services, and want "one gateway to unify them all",, am I correct?

Currently we *do* have plans for this but they aren't a priority. See https://github.com/kanidm/kanidm/issues/149

**Reply by coopbri** [reply]
Yes exactly. Thanks @Firstyear -- noted. No worries at all, easy enough to integrate dex for now (it's pretty lightweight) and switch later if/when it becomes a reality with kanidm. Cheers

**Reply by Firstyear** [reply]
Sounds good - sorry that we haven't got the feature today.

**Reply by coopbri** [reply]
No need! Complex project, plenty of amazing features as is.
